### PR TITLE
Noor hazim w4 js

### PR DESCRIPTION
**Pull request by noorhazim**

### CODE REVIEW

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js [hunk 1]
```diff
@@ -16,36 +16,43 @@ Complete the function called `sanitizeFruitBasket`:
...
-  });
-
-  test('should return a new array that does not include the unwanted `lemon`', () => {
-    // TODO replace next line with your code
-    expect(false).toBe(true);
-  });
-});
+    test('should take two parameters', () => {
+        // TODO replace next line with your code
+        expect(sanitizeFruitBasket(fruitBasket, 'lemon')).toHaveLength(4);
```
Comment (killoncio) [inline 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:42]: Here you are not checking how many parameters the function `sanitizeFruitBasket` takes, you are calling the function, and checking the length of the result of calling that function.

Moreover, as the test says, it should have a length of 2, since it takes two parameters, not of 4 :)

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js [hunk 2]
```diff
@@ -16,36 +16,43 @@ Complete the function called `sanitizeFruitBasket`:
...
+    });
+
+    test('should not modify the original `fruitBasket` array', () => {
+        // Save the original contents of the fruit basket
+        const originalFruitBasketContents = [...fruitBasket];
+        // TODO replace next line with your code
+        expect(originalFruitBasketContents).toHaveLength(7);
```
Comment (killoncio) [inline 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:49]: Here you are checking that `originalFruitBasketContents` array has a length of 7. But you don't know if the elements have changed, which would mean the original array has been modified.

Can you think of a way to confirm that the original array has not changed after being passed to the function?

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js [hunk 3]
```diff
@@ -16,36 +16,43 @@ Complete the function called `sanitizeFruitBasket`:
...
+    });
+
+    test('should return a new array that does not include the unwanted `lemon`', () => {
+        // TODO replace next line with your code
+        const result = sanitizeFruitBasket(fruitBasket, 'lemon');
+        const expected = ['apple', 'grapefruit', 'banana', 'watermelon'];
```
Comment (killoncio) [inline 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:55]: This is one way to confirm that the new array does not include `lemon`. But, in order to do that, you are creating a new array hardcoding all the values that are in the new array, which can be cumbersome if the array is longer.

Can you think of another way to check if an element is not in the array?

File: 1-JavaScript/Week4/homework/ex4-observable/ex4-observable.js [hunk 4]
```diff
@@ -14,16 +14,22 @@ Complete the `createObservable()` function as follows:
...
-  const subscribers = [];
-  return {
-    subscribe: function (subscriber) {
-      // TODO complete this function
-    },
-    notify: function (message) {
-      // TODO complete this function
-    },
-  };
+    const subscribers = [];
```
Comment (killoncio) [inline 1-JavaScript/Week4/homework/ex4-observable/ex4-observable.js:17]: The solution is correct, well done! 

Can you please remove the extra empty lines in the function? It helps to the readability of the code.

File: 1-JavaScript/Week4/homework/ex1-doubleEvenNumbers.test.js [hunk 5]
```diff
@@ -11,19 +11,19 @@ Let's rewrite it (or _refactor_ it, as experienced developers would call it):
...
-  // TODO rewrite the function body using `map` and `filter`.
-  const newNumbers = [];
-  for (let i = 0; i < numbers.length; i++) {
-    if (numbers[i] % 2 === 0) {
-      newNumbers.push(numbers[i] * 2);
-    }
-  }
-  return newNumbers;
+    // TODO rewrite the function body using `map` and `filter`.
+    const newNumbers = numbers.filter(number => {
```
Comment (killoncio) [inline 1-JavaScript/Week4/homework/ex1-doubleEvenNumbers.test.js:15]: Perfect!

File: 1-JavaScript/Week4/homework/ex2-mondaysWorth.test.js [hunk 6]
```diff
@@ -10,42 +10,47 @@ Full description atL https://github.com/HackYourFuture/Homework/tree/main/1-Java
...
 
 const hourlyRate = 25;
 
-function computeEarnings(/* TODO parameter(s) go here */) {
-  // TODO complete this function
+function computeEarnings(tasks, rate) {
+    const ratePerTask = tasks.map(function(element) {
+        return (element.duration) * rate / 60;
+    })
+    const earning = (prev, cur) => prev + cur;
```
Comment (killoncio) [inline 1-JavaScript/Week4/homework/ex2-mondaysWorth.test.js:37]: Well done! 

It's nice that you extract the reducer function outside and stored it as an element. I'd use a different name though, since `earning` might be misleading. Normally it's used the name `reducer` since it's what the function does https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js [hunk 7]
```diff
@@ -16,36 +16,43 @@ Complete the function called `sanitizeFruitBasket`:
...
+    'lemon',
+    'banana',
+    'watermelon',
+    'lemon',
 ];
 
 // ! Function under test
-function sanitizeFruitBasket(/* TODO parameter(s) go here */) {
-  // TODO complete this function
+function sanitizeFruitBasket(fruitBasket, outFruit) {
```
Comment (killoncio) [inline 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:29]: Nice!

Can you please remove the extra empty lines?

File: 1-JavaScript/Week3/homework/ex1-giveCompliment.js [hunk 8]
```diff
@@ -18,27 +18,30 @@ Full description at: https://github.com/HackYourFuture/Homework/tree/main/1-Java
    Use `console.log` each time to display the return value of the
    `giveCompliment` function to the console.
 -----------------------------------------------------------------------------*/
-function giveCompliment(/* TODO parameter(s) go here */) {
-  // TODO complete this function
+function giveCompliment(name) {
```
Comment (killoncio) [inline 1-JavaScript/Week3/homework/ex1-giveCompliment.js:21]: Since this PR is for the week 4 exercises, it would be good to remove anything that is not related to that, like the exercises of week 3.

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js (hunk 7 above)
Comment (noorhazim) [reply 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:29]: Sure, Done!

File: 1-JavaScript/Week4/homework/ex4-observable/ex4-observable.js (hunk 4 above)
Comment (noorhazim) [reply 1-JavaScript/Week4/homework/ex4-observable/ex4-observable.js:17]: Done!

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js (hunk 3 above)
Comment (noorhazim) [reply 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:55]: I use   expect(result).not.toContain('lemon') I think it works. do I need any other ideas?

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js (hunk 2 above)
Comment (noorhazim) [reply 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:49]: is using .toEqual(fruitBasket) good idea to check if it's changed or not?

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js (hunk 1 above)
Comment (noorhazim) [reply 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:42]: Done!

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js (hunk 2 above)
Comment (killoncio) [reply 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js]: Yes! Well done!

**Comment by killoncio** [comment]
Well done, Noor! Everything is good now!
